File: backend/server/app/pkg/logger/middleware.go
package logger

import (
	"log/slog"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/google/uuid"
	apptracing "github.com/woragis/backend/server/app/pkg/tracing"
)

// RequestIDMiddleware generates and adds a trace_id (request ID) to each request.
// The trace_id is added to the context and response headers for distributed tracing.
// This works with OpenTelemetry tracing - if a trace ID exists from OpenTelemetry,
// it will be used; otherwise a new one is generated.
func RequestIDMiddleware(log *slog.Logger) fiber.Handler {
	return func(c *fiber.Ctx) error {
		ctx := c.UserContext()
		
		// First, try to get trace ID from OpenTelemetry span context
		// (if tracing middleware ran first)
		traceID := apptracing.TraceIDFromContext(ctx)
		
		// If no trace ID from OpenTelemetry, check header
		if traceID == "" {
			traceID = c.Get("X-Trace-ID")
		}
		
		// If still no trace ID, generate new one
		if traceID == "" {
			traceID = uuid.New().String()
		}

		// Add trace_id to response header
		c.Set("X-Trace-ID", traceID)

		// Add trace_id to context for logging (preserves OpenTelemetry context)
		ctx = WithTraceID(ctx, traceID)
		c.SetUserContext(ctx)

		// Store trace_id in locals for easy access
		c.Locals("trace_id", traceID)

		return c.Next()
	}
}

// RequestLoggerMiddleware logs HTTP requests with structured fields.
// This should be used after RequestIDMiddleware to ensure trace_id is available.
func RequestLoggerMiddleware(log *slog.Logger) fiber.Handler {
	return func(c *fiber.Ctx) error {
		start := time.Now()

		// Process request
		err := c.Next()

		// Calculate duration
		duration := time.Since(start)

		// Get trace_id from context
		traceID := GetTraceID(c.UserContext())
		if traceID == "" {
			// Fallback to locals if not in context
			if id, ok := c.Locals("trace_id").(string); ok {
				traceID = id
			}
		}

		// Get user ID if available (from JWT middleware)
		userID := ""
		// Check auth middleware's user ID key
		if user, ok := c.Locals("auth.user_id").(string); ok {
			userID = user
		} else if userUUID, ok := c.Locals("auth.user_id").(uuid.UUID); ok {
			userID = userUUID.String()
		}

		// Build log attributes
		attrs := []slog.Attr{
			slog.String("method", c.Method()),
			slog.String("path", c.Path()),
			slog.String("ip", c.IP()),
			slog.Int("status", c.Response().StatusCode()),
			slog.Duration("duration", duration),
			slog.Int("bytes_in", len(c.Body())),
			slog.Int("bytes_out", len(c.Response().Body())),
		}

		if traceID != "" {
			attrs = append(attrs, slog.String("trace_id", traceID))
		}

		if userID != "" {
			attrs = append(attrs, slog.String("user_id", userID))
		}

		// Add query parameters if present
		if len(c.Queries()) > 0 {
			attrs = append(attrs, slog.String("query", c.OriginalURL()))
		}

		// Log based on status code
		ctx := c.UserContext()
		if traceID != "" {
			ctx = WithTraceID(ctx, traceID)
		}

		status := c.Response().StatusCode()
		switch {
		case status >= 500:
			log.LogAttrs(ctx, slog.LevelError, "http request", attrs...)
		case status >= 400:
			log.LogAttrs(ctx, slog.LevelWarn, "http request", attrs...)
		default:
			log.LogAttrs(ctx, slog.LevelInfo, "http request", attrs...)
		}

		return err
	}
}

File: backend/creative-service/app/middleware.py
"""
FastAPI middleware for structured logging and trace ID handling.
"""
import uuid
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware

from app.logger import get_logger, set_trace_id, get_trace_id

logger = get_logger()


class RequestIDMiddleware(BaseHTTPMiddleware):
    """Middleware to generate and propagate trace IDs (request IDs)."""
    
    async def dispatch(self, request: Request, call_next):
        # Check if trace_id already exists in header (for distributed tracing)
        trace_id = request.headers.get("X-Trace-ID")
        if not trace_id:
            # Generate new trace_id if not present
            trace_id = str(uuid.uuid4())
        
        # Set trace_id in context
        set_trace_id(trace_id)
        
        # Add trace_id to response header
        response = await call_next(request)
        response.headers["X-Trace-ID"] = trace_id
        
        return response


class RequestLoggerMiddleware(BaseHTTPMiddleware):
    """Middleware to log HTTP requests with structured fields."""
    
    async def dispatch(self, request: Request, call_next):
        import time
        start_time = time.time()
        
        # Get request body size (if available)
        body_size = 0
        if hasattr(request, "_body"):
            body_size = len(request._body) if request._body else 0
        
        # Process request
        response = await call_next(request)
        
        # Calculate duration
        duration = time.time() - start_time
        
        # Get trace_id
        trace_id = get_trace_id()
        
        # Build log attributes
        log_data = {
            "method": request.method,
            "path": request.url.path,
            "ip": request.client.host if request.client else None,
            "status": response.status_code,
            "duration_ms": round(duration * 1000, 2),
        }
        
        if body_size > 0:
            log_data["bytes_in"] = body_size
        
        if trace_id:
            log_data["trace_id"] = trace_id
        
        # Add query parameters if present
        if request.query_params:
            log_data["query"] = str(request.query_params)
        
        # Log based on status code
        status = response.status_code
        if status >= 500:
            logger.error("http request", **log_data)
        elif status >= 400:
            logger.warn("http request", **log_data)
        else:
            logger.info("http request", **log_data)
        
        return response
